TNL-4214: Add safe template linter

robrap opened the pull request:
Here are the direct results of my Hackathon project for a Safe Template Linter.

See [TNL-4214](https://openedx.atlassian.net/browse/TNL-4214) for many notes about enhancements that are needed.

Consider this both a very early draft review, as well as providing early access to the tool before it is ready to be merged.

Usage:
```
edxapp@precise64:~/edx-platform$ scripts/safe_template_checker.py
```
FYI: @cahrens @nedbat

nedbat commented:
How is it that this passes diff-quality pylint if there are no docstrings?? :)

nedbat commented on scripts/safe_template_checker.py:
```diff
@@ -0,0 +1,498 @@
+    try:
+        file_contents = input_file.read()
+    finally:
+        input_file.close()
```
Use a with-statement instead of try/finally.

robrap replied:
Agreed. When we last saw this together I was guessing this was the result of an improper refactor.

nedbat commented on scripts/safe_template_checker.py:
```diff
@@ -0,0 +1,498 @@
+        input_file.close()
+
+    if not file_contents:
+        return False
```
I'm not sure what this is for: the callers don't seem to use False, and it's confusing to return bool instead of an empty string.

robrap replied:
Agreed. When we last saw this together I was guessing this was the result of an improper refactor.

robrap replied:
Got it.  I did test the following:
```
alert('<script>alert("Hi!");</script>');
```
The browser dialog expects plain text and the entirety of the text is displayed.  I imagine the same thing will happen with this dialog.  Thanks.

nedbat commented on scripts/safe_template_checker.py:
```diff
@@ -0,0 +1,498 @@
+    def prepare_results(self, full_path, file_string, line_breaks):
+        self.full_path = full_path
+        start_index = self.expression['start_index']
+        self.start_line = _get_line_number(self, line_breaks, start_index)
```
All of this logic makes me think that lines 42-71 should be a class called IndexedSourceFile or something.

nedbat commented on scripts/safe_template_checker.py:
```diff
@@ -0,0 +1,498 @@
+        results.prepare_results(mako_template)
+
+    def _has_page_default(self, mako_template, results):
+        page_h_filter_regex = re.compile('<%page expression_filter=(?:"h"|\'h\')\s*/>')
```
All regexes should be written with raw strings: `r"..."`.

nedbat replied:
There's no advantage to compiling these before using them, and it forces you to come up with a name for them.

robrap commented:
@nedbat There is no new functionality.  I did some renaming.  I added the doc strings.  I also refactored to print by file, which is better because then the output is sorted by file path rather than template type.

Please either read doc strings and comment if you'd like, or provide a thumb if you think I should squash and merge and we'll get into that later.

robrap commented:
For reference, @nedbat gave me a thumb outside of this PR.  I am going to merge to prepare for the Safe Template Marathon.  This may be refactored into its own repo at a later date.
